### PR TITLE
Add #publication_date_for_slider for extracting date ranges

### PR DESCRIPTION
**Pull request by cbeer**

### CODE REVIEW

**Comment by cbeer** [comment]
I'll entertain arguments that I should also update the date parsing logic in this gem to use the upstream logic.

File: lib/stanford-mods/imprint.rb [hunk 1]
```diff
@@ -37,8 +37,44 @@ def display_str
...
+                          end
+
+          if date_elements.none?
+            []
+          elsif date_elements.find(&:start?) && date_elements.find(&:end?)
+            start_date = date_elements.find(&:start?)
+            end_date = date_elements.find(&:end?)
+
+            (start_date.as_range.min.year..end_date.as_range.max.year).to_a
+          elsif date_elements.find(&:start?)
```
Comment (jcoyne) [inline lib/stanford-mods/imprint.rb]: can you `elsif start_date = date_elements.find(&:start?)` here?

File: lib/stanford-mods/imprint.rb [hunk 2]
```diff
@@ -37,8 +37,44 @@ def display_str
...
+                            origin_info_node.as_object.first.key_dates.map(&:as_object).map(&:first)
+                          else
+                            date_field_keys.map do |date_field|
+                              next unless origin_info_node.respond_to?(date_field)
+                              date_elements = origin_info_node.send(date_field)
+                              date_elements.map(&:as_object).map(&:first) if date_elements.any?
+                            end.first
+                          end
+
+          if date_elements.none?
```
Comment (jcoyne) [inline lib/stanford-mods/imprint.rb]: I'd like to see this conditional extracted to a separate method.

File: lib/stanford-mods/imprint.rb [hunk 3]
```diff
@@ -37,8 +37,44 @@ def display_str
...
+
+            (start_date.as_range.min.year..end_date.as_range.max.year).to_a
+          elsif date_elements.find(&:start?)
+            start_date = date_elements.find(&:start?)
+
+            (start_date.as_range.min.year..Time.now.year).to_a
+          elsif date_elements.one?
+            date_elements.first.to_a.map(&:year)
+          else
+            date_elements.map { |v| v.to_a.map(&:year) }.flatten
```
Comment (jcoyne) [inline lib/stanford-mods/imprint.rb]: Can you add a comment here about what you are expecting `date_elements` to look like when we hit this branch?

File: lib/stanford-mods/imprint.rb [hunk 4]
```diff
@@ -37,8 +37,44 @@ def display_str
...
+          end
+        end.flatten
+      end
+
       private
 
+      def extract_year(el)
```
Comment (jcoyne) [inline lib/stanford-mods/imprint.rb]: Add a comment about the return value. Is it an Integer?

File: lib/stanford-mods/imprint.rb [hunk 5]
```diff
@@ -37,8 +37,44 @@ def display_str
         imprint_statements.join('; ') if imprint_statements.present?
       end
 
+      def publication_date_for_slider
```
Comment (jcoyne) [inline lib/stanford-mods/imprint.rb]: Please add a comment about what type(s) this method returns.

File: lib/stanford-mods/imprint.rb (hunk 1 above)
Comment (cbeer) [reply lib/stanford-mods/imprint.rb]: I could, but does it help readability? The list of date elements (oughn't..) be very long; for the most part, it's 1-3 items, so performance isn't a huge concern.

File: lib/stanford-mods/imprint.rb (hunk 3 above)
Comment (cbeer) [reply lib/stanford-mods/imprint.rb]: Literally anything else?

Comment (jcoyne) [reply lib/stanford-mods/imprint.rb]: Is that so?  It looks to me like you are expecting an enumerable that is populated with methods that respond to `#to_a`?

Comment (cbeer) [reply lib/stanford-mods/imprint.rb]: That's true -- but this method is producing that for us.

File: lib/stanford-mods/imprint.rb (hunk 1 above)
Comment (jcoyne) [reply lib/stanford-mods/imprint.rb]: Okay, and we're not going to run this code on tens of thousands of records at a time?

Comment (cbeer) [reply lib/stanford-mods/imprint.rb]: No, this code acts on single records.

**Pull request merged**